### PR TITLE
🪐 fix: Replace `$bitsAllSet` ACL Queries for Azure Cosmos DB Compatibility

### LINKED ISSUE #12729: [Bug]: `$bitsAllSet` queries in `aclEntry.ts` break Azure Cosmos DB for MongoDB deployments
### What happened?

The new ACL system introduced in v0.8.5 (`packages/data-schemas/src/methods/aclEntry.ts`) uses the MongoDB bitwise query operator `$bitsAllSet`, which is **not supported by Azure Cosmos DB for MongoDB API**. Microsoft’s own Mongo API 7.0 feature support matrix lists `$bitsAllSet` as unsupported.

The earlier FerretDB Compatibility PR (#11769, merged Feb 13) correctly replaced `$bitsAllSet` with application-layer equivalents in the old `PermissionService` code. However, PRs #11896 (System Grants) and #12354 (Per-Principal Config) introduced a brand-new `aclEntry.ts` that reintroduced `$bitsAllSet` in multiple query paths, undoing the earlier compatibility fix for read-side permission checks.

The current source also uses `$bit` in `modifyPermissionBits()`, but Microsoft docs list `$bit` as supported on Azure Cosmos DB for MongoDB 7.0. The proven Cosmos-specific blocker from observed failures is therefore `$bitsAllSet`, not `$bit`.

**Expected behavior:** Permission queries should work on Azure Cosmos DB for MongoDB without failing on unsupported bitwise query operators.

**Actual behavior:** Every call that reaches a `$bitsAllSet` permission query fails on Cosmos DB. Agents and prompts do not appear in the UI, and direct agent reads fall through to `403 Forbidden` / `Agent not found` behavior because permission evaluation fails.

### Confirmed affected code paths in `packages/data-schemas/src/methods/aclEntry.ts`

| Function | Operator | Usage |
|---|---|---|
| `hasPermission()` | `$bitsAllSet` (query) | `permBits: { $bitsAllSet: permissionBit }` |
| `findAccessibleResources()` | `$bitsAllSet` (query) | `permBits: { $bitsAllSet: requiredPermBit }` |
| `findPublicResourceIds()` | `$bitsAllSet` (query) | `permBits: { $bitsAllSet: requiredPermissions }` |
| `getSoleOwnedResourceIds()` | `$bitsAllSet` (query) × 2 | `permBits: { $bitsAllSet: PermissionBits.DELETE }` (both in owned-entry query and other-owners aggregation) |

### Related implementation detail

`modifyPermissionBits()` still uses `$bit` for updates. That may still matter for FerretDB or broader non-Mongo compatibility, but it is not the Cosmos-specific root cause established by the current logs and Microsoft support matrix.

### Prior art

PR #11769 (FerretDB Compatibility) already solved this exact problem for the _old_ code by replacing `$bitsAllSet` queries with application-layer bitwise filtering (read + check `(permBits & required) === required`). The same pattern should be applied to the new `aclEntry.ts` query methods.

### Related issues

- #10998 / Discussion #10999 - "Exploring Marketplace not showing shared agents" - another user on Azure Container Apps + Cosmos DB reported identical `$bitsAllSet` errors in December 2025. The response directed them to Discussion #9429 (index creation), but the root cause is unsupported query operators, not missing indexes.

Both verified on Azure Container Apps with Azure Cosmos DB for MongoDB (serverless, MongoDB API 7.0).

External confirmation:
- Azure Cosmos DB for MongoDB 7.0 feature support matrix: `$bitsAllSet` = `✖️ No`, `$bit` = `✅ Yes`
- https://learn.microsoft.com/azure/cosmos-db/mongodb/feature-support-70#bitwise-update-operator

## Suggested Fix

Apply the same pattern established in PR #11769 to the new `aclEntry.ts` query methods:

**For `$bitsAllSet` queries** - Replace database-side bitwise filtering with application-layer filtering:
```typescript
// Before (breaks on Cosmos DB / FerretDB):
const entries = await AclEntry.find({
  $or: principalsQuery,
  resourceType,
  permBits: { $bitsAllSet: requiredPermBit },
}).distinct('resourceId');

// After (compatible with all MongoDB-API databases):
const entries = await AclEntry.find({
  $or: principalsQuery,
  resourceType,
}).lean();
const filtered = entries.filter(e => (e.permBits & requiredPermBit) === requiredPermBit);
return [...new Set(filtered.map(e => e.resourceId))];
```

**Performance note:** The application-layer filtering adds marginal overhead (fetching slightly more documents than needed), but for typical ACL entry counts this is negligible.

### Version Information

v0.8.4 and v0.8.5-rc1

### Steps to Reproduce

1. Deploy LibreChat v0.8.5-rc1 (or v0.8.4) with Azure Cosmos DB for MongoDB as the database backend (serverless or provisioned - both use the same MongoDB API)
2. Log in and navigate to the Agents page
3. Observe that no agents appear (even if agents were previously created)
4. Check container logs - you will see repeated errors:
   ```
   [PermissionService.findAccessibleResources] Error: $bitsAllSet not supported
   [PermissionService.findPubliclyAccessibleResources] Error: $bitsAllSet not supported
   ```

The errors occur on every page load that touches agents, prompts, or any ACL-protected resource. Core chat functionality (sending messages) still works because it doesn't go through the same ACL permission query path.

### What browsers are you seeing the problem on?

Microsoft Edge

### Relevant log output

```shell
2026-04-18T05:49:07.123456789Z 2026-04-18 05:49:07 error: [PermissionService.findAccessibleResources] Error: $bitsAllSet not supported
2026-04-18T05:49:07.234567890Z 2026-04-18 05:49:07 error: [PermissionService.findAccessibleResources] Error: $bitsAllSet not supported
2026-04-18T05:49:07.345678901Z 2026-04-18 05:49:07 error: [PermissionService.findAccessibleResources] Error: $bitsAllSet not supported
2026-04-18T05:49:07.456789012Z 2026-04-18 05:49:07 error: [PermissionService.findAccessibleResources] Error: $bitsAllSet not supported
2026-04-18T05:49:12.567890123Z 2026-04-18 05:49:12 error: [PermissionService.findAccessibleResources] Error: $bitsAllSet not supported
2026-04-18T05:49:12.678901234Z 2026-04-18 05:49:12 error: [PermissionService.findPubliclyAccessibleResources] Error: $bitsAllSet not supported
2026-04-18T05:49:13.789012345Z 2026-04-18 05:49:13 error: [PermissionService.findAccessibleResources] Error: $bitsAllSet not supported
2026-04-18T05:49:13.890123456Z 2026-04-18 05:49:13 error: [PermissionService.findPubliclyAccessibleResources] Error: $bitsAllSet not supported


These errors repeat on every page navigation that involves agents or prompts. The volume is proportional to the number of resource types checked per request (agents, prompts, MCP servers, etc.).
```

### Screenshots

_No response_

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by danny-avila**
## Summary

- Replaces every ACL read that used `$bitsAllSet` with application-layer bitwise filtering, restoring permission checks on Azure Cosmos DB for MongoDB (which does not implement `$bitsAllSet`).
- Touches `packages/data-schemas/src/methods/aclEntry.ts` (5 sites: `hasPermission`, `findAccessibleResources`, `findPublicResourceIds`, and two sites in `getSoleOwnedResourceIds`), plus the two remaining inlined queries in `packages/api/src/acl/accessControlService.ts` and `packages/api/src/apiKeys/permissions.ts`.
- Adds targeted unit tests covering exact-bit / superset / subset / disjoint cases for each rewritten method, including full coverage for `getSoleOwnedResourceIds` which was previously untested.

## Why

Per [Microsoft's Cosmos DB for MongoDB 7.0 feature support matrix](https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/feature-support-70), `$bitsAllSet` is **not** supported. Every ACL read path threw on Cosmos, which translated to empty agent/prompt lists, 403s on direct resource reads, and broken marketplace rendering (see [#12729](https://github.com/danny-avila/LibreChat/issues/12729) and prior report [#10998](https://github.com/danny-avila/LibreChat/issues/10998)). The `$bit` update operator used in `modifyPermissionBits` **is** supported and is left untouched.

The bitwise filter now runs in Node after a bounded `find()`, matching the pattern already validated for FerretDB compatibility (`packages/data-schemas/misc/ferretdb/aclBitops.ferretdb.spec.ts`). Fetched document counts are bounded by the number of ACL entries for a given principal / public principal / single resource, so the space-time tradeoff is negligible at real-world ACL scale.

Fixes #12729.

## Changes

1. **`packages/data-schemas/src/methods/aclEntry.ts`** — five `$bitsAllSet` queries replaced with `find().lean()` + JS-side `(permBits & bit) === bit`. The aggregation pipeline in `getSoleOwnedResourceIds` is rewritten as a second `find()` + `Set`-based grouping.
2. **`packages/api/src/acl/accessControlService.ts`** — `findPubliclyAccessibleResources` now delegates to the already-fixed `_dbMethods.findPublicResourceIds` instead of duplicating the query.
3. **`packages/api/src/apiKeys/permissions.ts`** — `enrichRemoteAgentPrincipals` projects `permBits` through the aggregation and filters `PermissionBits.SHARE` in application code.

## Test plan

- [x] `packages/data-schemas`: all 1296 existing tests pass, plus 13 new cases under `Application-layer bitwise filtering (Cosmos DB compatibility)` in `aclEntry.spec.ts`.
- [x] `packages/api`: full ACL and permissions suites pass (`src/acl`, `src/apiKeys` — 56 tests).
- [x] `api/server/services/PermissionService.spec.js`: full suite passes (61 tests).
- [x] Related downstream: `agent.spec.ts` (111) and `prompt.spec.ts` (10) — both exercise `getSoleOwnedResourceIds` via dependency injection.
- [x] `tsc --noEmit` clean on both `packages/data-schemas` and `packages/api`.
- [x] `eslint` clean on all four changed files.
- [ ] Deploy against a real Azure Cosmos DB for MongoDB API 7.0 instance and verify agent/prompt list rendering end-to-end (requires reporter's environment).
- [ ] Optional: run the FerretDB integration suite (`packages/data-schemas/misc/ferretdb/aclBitops.ferretdb.spec.ts`) to confirm parity.

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
Thanks @chatgpt-codex-connector — acknowledged, no findings to address.

@codex review

**Comment by danny-avila** [comment]
@chatgpt-codex-connector thanks — good catch on the misleading test name. Addressed in [a1d4dc7](https://github.com/danny-avila/LibreChat/commit/a1d4dc722): renamed the single-entry case to reflect what it actually verifies, and added a genuine deduplication test that bypasses `grantPermission`'s upsert via `AclEntry.create` so the application-layer dedup `Map` is actually exercised.

(Note: the commit you mentioned on branch `work` (`ceedb1e`) isn't present on this repo — only treating your in-repo suggestions as ground truth.)

@codex review

**Comment by danny-avila** [comment]
Thanks @chatgpt-codex-connector. Running one more pass to confirm there's nothing left.

@codex review

**Comment by danny-avila** [comment]
Thanks @chatgpt-codex-connector. Two clean passes confirm nothing left to address — not pursuing the optional deeper dive since the existing coverage plus the FerretDB compatibility spec (`packages/data-schemas/misc/ferretdb/aclBitops.ferretdb.spec.ts`) already exercises Cosmos-equivalent parity.

Closing the review loop here. All 10 CI checks green on [a1d4dc7](https://github.com/danny-avila/LibreChat/commit/a1d4dc722). 🎉

**Comment by danny-avila** [comment]
Addressed all three findings from the comprehensive audit:

**Finding 1 (MAJOR)** — Missing test for SHARE bit filter in `enrichRemoteAgentPrincipals`. Added [packages/api/src/apiKeys/permissions.spec.ts](https://github.com/danny-avila/LibreChat/blob/c12fb429e/packages/api/src/apiKeys/permissions.spec.ts) with 6 dependency-injected unit tests covering: SHARE present / absent / zero-bit / superset, missing user info, idempotency against already-listed principals, and a regression guard that the `$match` stage no longer contains a `permBits` filter. ([6fe4a1c](https://github.com/danny-avila/LibreChat/commit/6fe4a1cda))

**Finding 2 (MINOR)** — DRY violation. Extracted the duplicated filter+dedup loop in `aclEntry.ts` into a file-local `filterByBitsAndDedup(entries, requiredBits)` helper; `findAccessibleResources` and `findPublicResourceIds` now both delegate. Pure refactor, tests unchanged. ([75cd652](https://github.com/danny-avila/LibreChat/commit/75cd652b6))

**Finding 3 (NIT)** — Stale `$bitsAllSet` references in `packages/data-schemas/misc/ferretdb/aclBitops.ferretdb.spec.ts`. Renamed the describe block to "Bitwise permission queries" and updated the file header comment to describe the application-layer behavior being validated. ([c12fb42](https://github.com/danny-avila/LibreChat/commit/c12fb429e))

Full suite re-verified: 1297 data-schemas tests, 62 `packages/api` ACL/apiKeys tests (including the 6 new `enrichRemoteAgentPrincipals` cases), 61 legacy `PermissionService` tests — all pass. `tsc --noEmit` + `eslint` clean.

**Comment by danny-avila** [comment]
Thanks for the follow-up audit. The remaining NIT (`permBits: undefined` test) is noted — skipping since the reviewer's own verdict is "safe to ignore": Mongoose schema validation (default `1`, `type: Number`) prevents `undefined` from ever reaching that branch in production, and the existing cast-then-mask chain (`NaN & SHARE === 0`) is already safe by construction. Adding a test for a state the schema excludes would contradict the "don't validate for scenarios that can't happen" guideline in `CLAUDE.md`.

All 3 original findings resolved, CI all green. I'll consider this review loop closed unless codex surfaces anything new.

**Comment by danny-avila** [comment]
Pushed the `$in`-based optimization plus a parity+performance spec. Two commits:

**[b22df71](https://github.com/danny-avila/LibreChat/commit/b22df716a)** — `perf: push permission-mask filter back to the database via `$in``
- Added memoized `permissionBitSupersets(requiredBits)` helper (exported from `@librechat/data-schemas`) that expands a required bitmask into the list of all permBits values whose bits are a superset. For the current 4-bit `PermissionBits` enum this is at most 16 values (8 for a single bit like SHARE).
- Replaced every `permBits: { $bitsAllSet: X }` with `permBits: { $in: permissionBitSupersets(X) }`. `$in` is indexable AND supported by Cosmos DB for MongoDB, so the filter runs server-side again.
- `hasPermission` back to `findOne` + short-circuit, `findAccessibleResources` / `findPublicResourceIds` back to `.distinct('resourceId')`, `getSoleOwnedResourceIds` back to `$match` + `$group` aggregation, `enrichRemoteAgentPrincipals` filter back inside `$match`.
- Updated the `enrichRemoteAgentPrincipals` test suite: split into "pipeline shape" tests (regression guard that `$match.permBits` is `$in` of the correct supersets, not `$bitsAllSet`) and "enrichment behavior" tests (mock returns pre-filtered rows → verify mapping).

**[a75f122](https://github.com/danny-avila/LibreChat/commit/a75f12219)** — `test: add `$bitsAllSet` vs `$in` parity + perf spec`
- New `aclEntry.parity.spec.ts` runs the legacy `$bitsAllSet` query and the current `$in` query against the same `mongodb-memory-server` fixture for every affected method across 7 meaningful permBits combinations. **All 14 parity checks pass — identical output sets.**
- Logs median wall-clock time over 20 runs on an 800-entry fixture with a loose 3× regression guard:
  - `findAccessibleResources` — legacy 1.05 ms vs current 1.07 ms
  - `findPublicResourceIds` — legacy 1.10 ms vs current 1.05 ms

### Re-verification
- **Data-schemas: 1311 tests pass** (up from 1297, +14 parity cases)
- **packages/api ACL/apiKeys: 63 tests pass** (up from 62)
- **Legacy PermissionService: 61 tests pass**
- `tsc --noEmit` + `eslint` clean

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
Thanks @chatgpt-codex-connector. One more pass to confirm.

@codex review

**Comment by danny-avila** [comment]
Audited the two parallel reviews and landed three hardening commits. Findings that were both valid and in-scope resolved; defensive-only suggestions where the reviewer themselves noted 'not a current bug' still landed when the fix was trivial.

**[754ade8](https://github.com/danny-avila/LibreChat/commit/754ade8e0)** — `hardening: freeze permissionBitSupersets cache + enum-shape guard`
- Review 1 F1 / Review 2 F1 (MINOR, 72-75): Cached arrays now `Object.freeze`d and returned as `readonly number[]`. Any mutation attempt now throws `TypeError` instead of silently corrupting the cache.
- Review 1 F2 / Review 2 F3 (MINOR/NIT, 65-85): Module-load guard `if (MAX_PERM_BITS === 0) throw`. If `PermissionBits` is ever refactored to a `const` object or string enum, the guard fails loudly instead of silently denying every permission check.
- Review 2 F4 (NIT, 95): Collapsed 4 duplicated JSDoc lines into a single `{@link permissionBitSupersets}` reference on each method.

**[114559b](https://github.com/danny-avila/LibreChat/commit/114559bce)** — `test: add focused unit tests for permissionBitSupersets`
- Review 1 F4 / Review 2 F2 (NIT/MINOR, 60): 6 new unit tests directly exercising the helper — boundary cases (`0` → all 16, `15` → `[15]`), universal superset invariant, full 0..15 parity against the `$bitsAllSet` definition, memoization identity, and the `TypeError` on mutation attempts.

**[21852ac](https://github.com/danny-avila/LibreChat/commit/21852acc5)** — `test: tighten parity perf guard and document fixture constants`
- Review 2 F5 (NIT, 80): Replaced `legacyMs * 3 + 50` with `Math.max(legacyMs * 5, 50)`. The additive form was dominated by `+50` at sub-ms latencies and would have let a 50× regression pass silently.
- Review 1 F3 (NIT, 80): Inline comments explaining why `FIXTURE_SIZE = 800` and `PERF_ITERATIONS = 20`.

### Findings deliberately skipped (with rationale)
- **`Record<string, unknown>` in test helpers** (both reviews, NIT, 55-65): Both reviewers themselves noted test code flexibility makes this acceptable. AGENTS.md pragmatism clause applies.
- **`hasPermission(..., 0)` not validated** (Review 2, NIT, 90): Explicitly pre-existing behavior, not introduced by this PR. Validation already exists one layer up at `AccessControlService.checkPermission`.
- **Inline `type` in value import** (Review 2, NIT, 95): Explicitly pre-existing convention — every import in `methods/index.ts` uses this form. Not PR scope.
- **Redundant `await`** (Review 2, NIT, 90): Reviewer themselves concluded "No action needed".

### Re-verification
- **Data-schemas: 1317 tests pass** (up from 1311 — 6 new permissionBitSupersets unit tests)
- **packages/api ACL/apiKeys: 63 tests pass**
- **Legacy PermissionService: 61 tests pass**
- `tsc --noEmit` + `eslint` clean

**Comment by danny-avila** [comment]
Follow-up audit addressed. Two commits, covering the one strong finding and three cosmetic ones:

**[8f7e8bf](https://github.com/danny-avila/LibreChat/commit/8f7e8bf1d)** — `chore: remove stale perf-guard comment, hoist rationale to describe`
- **Finding 1 (NIT, 95)** — valid and correct. Commit [a75f122](https://github.com/danny-avila/LibreChat/commit/a75f12219) had left a "Sanity check: A 3x multiplier" comment block, and [21852ac](https://github.com/danny-avila/LibreChat/commit/21852acc5) layered a second "Multiplicative-dominant" block right below it. Both blocks coexisted above the first `expect` and contradicted each other. Deleted the stale block, removed the now-redundant copy from the second perf test, and lifted the single rationale to the enclosing `describe('performance')`. Each perf test is now a bare `expect` — rationale lives once at the scope it applies to.
- **Finding 2 (NIT, 85-90)** — same fix. The duplicated rationale is now gone.

**[9464d8e](https://github.com/danny-avila/LibreChat/commit/9464d8e37)** — `docs: sharpen MAX_PERM_BITS guard message + test-sort consistency`
- **Finding C-1 (NIT, 90)** — `MAX_PERM_BITS === 0` error now names the specific refactors (const object / string enum / all-string shape) and gives two concrete remediation paths.
- **Finding R-1 (NIT, 55)** — `requiredBits=15` unit test now applies the same `[...result].sort((a, b) => a - b)` normalization as the other tests. The helper's JSDoc doesn't promise order, so sorting before set-equality comparison is the right defensive pattern.

### Findings deliberately skipped
- **R-2 / strict-mode dependency of the freeze-mutation test** — Review 1 itself noted "No action needed. Modern ts-jest/swc always emit strict." The test name already documents the assumption.

### Re-verification
- Data-schemas: 1317 tests pass
- `tsc --noEmit` + `eslint` clean
- Parity perf test still green

This should be the last round.

**Comment by danny-avila** [comment]
@codex review

File: packages/data-schemas/src/methods/aclEntry.ts [hunk 1]
```diff
@@ -10,6 +10,62 @@ import type {
...
+const supersetCache = new Map<number, readonly number[]>();
+
+export function permissionBitSupersets(requiredBits: number): readonly number[] {
+  const cached = supersetCache.get(requiredBits);
+  if (cached !== undefined) {
+    return cached;
+  }
+  const supersets: number[] = [];
+  for (let v = 0; v <= MAX_PERM_BITS; v++) {
+    if ((v & requiredBits) === requiredBits) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/data-schemas/src/methods/aclEntry.ts:64]: Valid and important finding — thanks. Fixed in [595935c8](https://github.com/danny-avila/LibreChat/commit/595935c84): added `min: 0`, `max: MAX_PERM_BITS`, and an `isInteger` validator to the `permBits` field. When a new bit is added to `PermissionBits`, `MAX_PERM_BITS` auto-expands, so the read-side enumeration and the write-side bound stay in sync. 4 new tests verify the schema rejects out-of-range / negative / non-integer values. Current write paths only produce values in `[0, 15]` via enums, so no live data is affected, but the divergence was reachable without the bound.

**Comment by danny-avila** [comment]
@codex review

File: packages/data-schemas/src/methods/aclEntry.ts [hunk 2]
```diff
@@ -10,6 +10,68 @@ import type {
...
+      supersets.push(v);
+    }
+  }
+  /**
+   * Freeze the cached array so a future caller cannot mutate the shared cache
+   * entry. All current callers only forward this to Mongoose's `$in`, which
+   * does not mutate — freezing is cheap and prevents silent corruption.
+   */
+  const frozen = Object.freeze(supersets);
+  supersetCache.set(requiredBits, frozen);
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/data-schemas/src/methods/aclEntry.ts:74]: Valid and important — this was a real DoS vector. Fixed in [0a0aa17](https://github.com/danny-avila/LibreChat/commit/0a0aa175b): `permissionBitSupersets` now rejects any `requiredBits` that isn't an integer, is negative, or has bits set above `MAX_PERM_BITS`, returning a shared frozen empty array WITHOUT caching. An empty `$in` list correctly matches zero rows, which is also semantically correct — rows cannot satisfy bits the system does not support.

Five new tests cover the rejection path including a 2000-unique-value cache-growth probe that asserts shared-instance identity across rejected inputs (proving nothing was cached). Meanwhile a regression test confirms legitimate in-range inputs still memoize normally.

**Comment by danny-avila** [comment]
Audited the latest comprehensive review plus the codex P2 DoS finding. Both resolved across two commits:

**[75ed77f](https://github.com/danny-avila/LibreChat/commit/75ed77f14)** — `refactor: hoist MAX_PERM_BITS + enum-shape/ceiling guards to shared util`
- **F1 (MINOR, 55)** — DRY violation. `MAX_PERM_BITS` was computed identically in `methods/aclEntry.ts` and `schema/aclEntry.ts`. Extracted to `packages/data-schemas/src/common/permissions.ts` and exported from the package barrel.
- **F2 (MINOR, 35)** — Schema-side `MAX_PERM_BITS === 0` guard missing. Now applied in a single place (the shared util), so both sides get it automatically.
- **F5 (NIT, 30)** — Circuit breaker for `MAX_PERM_BITS > 255`. At >8 bits the `$in` enumeration would emit >256 values and degrade query plans; fail loudly at module load instead.

**[0a0aa17](https://github.com/danny-avila/LibreChat/commit/0a0aa175b)** — `fix: reject out-of-range requiredBits without caching (DoS fix)`
- **Codex P2 / Audit F4 (MINOR, 50)** — User-controlled input path. `api/server/controllers/agents/v1.js` parses `req.query.requiredPermission` via `parseInt` and forwards it unvalidated, so an attacker could flood the memoized `supersetCache` with unique integers and grow memory unboundedly. Fix: reject non-integers, negatives, and inputs with bits above `MAX_PERM_BITS` — return a single shared frozen empty array and skip caching entirely. Five tests cover the rejection path, including a 2000-unique-value cache-growth probe.

### Findings deliberately skipped
- **F3 (MINOR, 40) — no migration check for pre-existing out-of-range rows:** Current code paths only produce `permBits` values in `[0, 15]` via the `PermissionBits`/`RoleBits` enums, and the PR description already calls this out. A startup-time scan `countDocuments({ permBits: { $gt: MAX_PERM_BITS } })` adds intrusive DB cost to boot for a risk the reviewer themselves rates at confidence 40. If a deployment later discovers rogue rows, they'll surface as unexpected 403s and can be remediated with a one-off migration script — which is cheaper than paying the boot cost on every startup for every deployment.
- **F6 (NIT, 25) — `Record<string, unknown>` in test helpers:** Skipped in prior rounds with reviewer agreement that test code flexibility applies.
- **F7 (NIT, 15) — `console.log` in perf tests:** Reviewer themselves labels this as intentional diagnostic output.

### Re-verification
- Data-schemas: 1326 tests pass (up from 1321)
- packages/api ACL/apiKeys: 63 pass
- Legacy PermissionService: 61 pass
- `tsc --noEmit` + `eslint` clean
